### PR TITLE
Feature/ab2d 2176 write contract mappings

### PR DESCRIPTION
**Pull request by wnyffenegger**
**JIRA Tickets:**

[AB2D-2176](https://jira.cms.gov/browse/AB2D-2176) - Write Contract Mappings

***Related Tickets***

[AB2D-2110](https://jira.cms.gov/browse/AB2D-2110) - DB Schema Updates
[AB2D-2146](https://jira.cms.gov/browse/AB2D-2146) - Coverage Mapping CRUD Service
[AB2D-2160](https://jira.cms.gov/browse/AB2D-2160) - Coverage Service Bulk Reads and Writes

### What Does This PR Do?

This PR

1. Implements saving beneficiary mapping of contract membership to our database but does not address pulling that information from our database
2. Creates a service and framework for discovering coverage membership data that is either missing, stale, or manually requested.
3. Executes coverage membership searches on demand with load balancing
4. Handles failed jobs gracefully allowing multiple attempts and notifying on failure
5. Gracefully handles shut downs making sure jobs do not get stuck (running longer than x hours)
6. Provides code for detecting stuck jobs and adding them back into the list of queued jobs

#### Major Changes

1. PatientContractCallable -> CoverageMappingCallable: create new callable that handles querying BFD for membership (needs to be updated to match updates to PatientContractCallable)
2. CoverageSearch entity and table in the database: list of queued
3. Database locking: locking the coverage search table so that only one worker can access the database table at a time to remove a row.
4. CoverageProcessor service: service for processing requests to look up coverage information. Automatically scans for searches in the CoverageSearch table.
5. CoverageMappingProcessor: implementation of coverage service that has a thread pool and manages load between the workers based on database insertion speed and configured max number of threads.
6. ThreadPool handling: add spring config parameters that let you control the thread pool for coverage mapping and control the checks for stale and stuck jobs.

#### Major Concerns
1. Database locking code in [ContractSearchLock](worker/src/main/java/gov/cms/ab2d/worker/processor/domainmodel/ContractSearchLock.java) which has the greatest potential to cause deadlocks
2. Synchronization code in [CoverageProcessorImpl](worker/src/main/java/gov/cms/ab2d/worker/processor/CoverageProcessorImpl.java) and potential for deadlocks
3. Shutdown code in [CoverageProcessorImpl](worker/src/main/java/gov/cms/ab2d/worker/processor/CoverageProcessorImpl.java) with potential for deadlocks and also need to make sure IN_PROGRESS jobs don't get stuck in that state even if they are shutdown.
4. New properties in application.properties and edge cases where they may not be set right
5. Edge cases that may occur when multiple workers are attempting to do work at the same time

### Database Changes

1. CoverageSearch entity and table added

### Limitations

1. Does not integrate with current EOB search at all
2. Cannot be run automatically, testing must be done manually
3. Does not have way to queue CoveragePeriods which are used to define work

### Security Implications

- [ ] This PR adds new software dependencies
- [ ] This PR modifies or invalidates our security controls
- [x] This PR stores or transmits data that was not stored or transmitted before

- [ ] This PR requires additional review of its security implications for other reasons

### Submitter Checklist

- [ ] This PR includes any required documentation changes, (`README`, changelog / release notes, website).
- [ ] All tech debt and/or shortcomings introduced by this PR are detailed in `TODO` and/or `FIXME` comments.
- [x] Code checked for PHI/PII exposure

### CODE REVIEW

File: common/src/main/java/gov/cms/ab2d/common/service/CoverageServiceImpl.java [hunk 1]
```diff
@@ -353,34 +378,110 @@ public CoverageSearchDiff searchDiff(int periodId) {
...
+            return Optional.of(updateStatus(period, description, JobStatus.SUBMITTED));
+        }
+
+        return Optional.empty();
+    }
+
+    @Override
+    public Optional<CoverageMapping> startSearch(CoverageSearch submittedSearch, String description) {
+
+        if (submittedSearch == null) {
```
Comment (nflearl) [inline common/src/main/java/gov/cms/ab2d/common/service/CoverageServiceImpl.java:472]: Already consistent with the above 2 suggestions!

File: common/src/test/java/gov/cms/ab2d/common/service/CacheServiceImplTest.java [hunk 2]
```diff
@@ -76,6 +74,13 @@ void setUp() {
         dataSetup.createCoverage(coveragePeriod, coverageSearchEvent, DataSetup.createBeneId());
     }
 
+    @AfterEach
+    void teardDown() {
+        coverageRepo.deleteAll();
```
Comment (nflearl) [inline common/src/test/java/gov/cms/ab2d/common/service/CacheServiceImplTest.java:79]: Bravo!  It was a real red flag to me to see deleteAll() in setup methods.  It makes it very hard to track down where the leaks actually are.

File: worker/src/main/java/gov/cms/ab2d/worker/processor/domainmodel/ContractSearchLock.java [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+ * Used to lock the database so that only one application or thread can retrieve a value from
+ * the database at a time since it also deletes the row. A thread can retrieve its own lock.
+ *
+ * This is based on Spring Integration and relies on the int_lock table. This and related tables
+ * should be created by Spring Integration on worker startup, although not necessarily before tests
+ * are run so if you don't have these tables, run worker first. The lock works by putting an
+ * entry in the int_lock table (that expires after an amount of time). When the lock is released,
+ * the entry is deleted. This allows different instances of worker to not step on each other.
+ */
+public class ContractSearchLock {
```
Comment (nflearl) [inline worker/src/main/java/gov/cms/ab2d/worker/processor/domainmodel/ContractSearchLock.java:26]: Note to self - circle back on this concept.

File: worker/src/test/java/gov/cms/ab2d/worker/processor/CoverageProcessorIntTest.java [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+
+        assertEquals(40000, coverageRepo.count());
+        assertEquals(4, coverageRepo.findAll()
+                .stream().map(c -> c.getCoveragePeriod().getMonth())
+                .collect(toSet()).size());
+    }
+
+    private void sleep(int seconds) {
+        try {
+            Thread.sleep(seconds * 1000);
```
Comment (nflearl) [inline worker/src/test/java/gov/cms/ab2d/worker/processor/CoverageProcessorIntTest.java:92]: 1000L will get rid of the Code Scanning warning.

**Review by nflearl — CHANGES_REQUESTED** [review]
This clearly was a massive amount of effort and overall, it looks to be fairly well done.

Hopefully my fresh set of eyes can help with some clarity improvements.

File: common/src/main/java/gov/cms/ab2d/common/service/CoverageServiceImpl.java [hunk 5]
```diff
@@ -1,13 +1,17 @@
 package gov.cms.ab2d.common.service;
+
```
[the remark this answers is not in the document]
Comment (wnyffenegger) [reply common/src/main/java/gov/cms/ab2d/common/service/CoverageServiceImpl.java:2]: Which methods should be leaving? I can move to a repository class fairly easily.

File: common/src/main/java/gov/cms/ab2d/common/service/CoverageServiceImpl.java [hunk 6]
```diff
@@ -163,11 +174,25 @@ private void insertBatch(int periodId, long searchEventId, List<String> batch) {
...
+     * @param offset offset into the past 0 is last search done successfully, 1 is search before that, etc.
+     */
+    private void deletePreviousSearch(int periodId, int offset) {
         CoveragePeriod period = findCoveragePeriod(periodId);
-        Optional<CoverageSearchEvent> searchEvent = coverageSearchEventRepo.findSearchWithOffset(periodId, 1);
-
+        Optional<CoverageSearchEvent> searchEvent = coverageSearchEventRepo
+                .findSearchEventWithOffset(periodId, JobStatus.IN_PROGRESS.name(), offset);
 
+        // Only delete previous search if a previous search exists.
```
[the remark this answers is not in the document]
Comment (wnyffenegger) [reply common/src/main/java/gov/cms/ab2d/common/service/CoverageServiceImpl.java:195]: @nflearl this was kept as plain jdbc because it is incredibly slow via hibernate.

File: common/src/main/java/gov/cms/ab2d/common/service/CoverageService.java [hunk 7]
```diff
@@ -104,23 +108,75 @@
...
+
     /**
      * Change a coverage search to {@link JobStatus#SUBMITTED} and log an event.
      * @param periodId unique id of a coverage search
      * @param description reason or explanation for change
      * @return resulting coverage search event
      * @throws InvalidJobStateTransition if job is {@link JobStatus#IN_PROGRESS} already
      */
-    CoverageSearchEvent submitCoverageSearch(int periodId, String description);
+    Optional<CoverageSearchEvent> submitSearch(int periodId, String description);
```
[the remark this answers is not in the document]
Comment (wnyffenegger) [reply common/src/main/java/gov/cms/ab2d/common/service/CoverageService.java:140]: Most of these events are necessary for things that have not been implemented yet. The only potential exceptions are submitSearch, prioritizeSearch, and vacuumCoverage. I already removed vacuumCoverage.

File: common/src/main/java/gov/cms/ab2d/common/repository/CoverageSearchEventRepository.java [hunk 8]
```diff
@@ -17,15 +19,23 @@
...
+            "   WHERE ebc.bene_coverage_period_id = :periodId AND ebc.new_status = :status" +
             "   ORDER BY ebc.created DESC LIMIT 1 OFFSET :offset",
             nativeQuery = true)
-    Optional<CoverageSearchEvent> findSearchWithOffset(int searchId, int offset);
+    Optional<CoverageSearchEvent> findSearchEventWithOffset(int periodId, String status, int offset);
+
+    @Query(value = "SELECT * FROM ( " +
+            "                  SELECT DISTINCT ON (cov_event.bene_coverage_period_id) cov_event.* " +
+            "                  FROM event_bene_coverage_search_status_change cov_event " +
+            "                  ORDER BY cov_event.bene_coverage_period_id, cov_event.created DESC " +
```
[the remark this answers is not in the document]
Comment (wnyffenegger) [reply common/src/main/java/gov/cms/ab2d/common/repository/CoverageSearchEventRepository.java:45]: Commented on it, not sure if it is clearer.

**Comment by sb-benohe** [comment]
SIA was Approved by ISSO

File: common/src/main/java/gov/cms/ab2d/common/service/CoverageServiceImpl.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (nflearl) [reply common/src/main/java/gov/cms/ab2d/common/service/CoverageServiceImpl.java]: I'm fine with plain JDBC in this case as long as the actual JDBC resides in a Repository class instead of a Service class.

File: worker/src/main/java/gov/cms/ab2d/worker/processor/CoverageMappingCallable.java [hunk 9]
```diff
@@ -0,0 +1,296 @@
...
+
+@Slf4j
+public class CoverageMappingCallable implements Callable<CoverageMapping> {
+
+    static final String BENEFICIARY_ID = "https://bluebutton.cms.gov/resources/variables/bene_id";
+    public static final String CURRENCY_IDENTIFIER =
+            "https://bluebutton.cms.gov/resources/codesystem/identifier-currency";
+    public static final String CURRENT_MBI = "current";
+    public static final String HISTORIC_MBI = "historic";
+    public static final String MBI_ID = "http://hl7.org/fhir/sid/us-mbi";
```
Comment (nflearl) [inline worker/src/main/java/gov/cms/ab2d/worker/processor/CoverageMappingCallable.java:27]: These are duplicate with those in PatientContraceCallable.  Just move them to something like a MbiSupport class.

File: worker/src/main/java/gov/cms/ab2d/worker/processor/CoverageMappingCallable.java [hunk 10]
```diff
@@ -0,0 +1,296 @@
...
+    public CoverageMapping call() {
+
+        int month = coverageMapping.getPeriod().getMonth();
+        String contractNumber = coverageMapping.getContract().getContractNumber();
+
+        final Set<Identifiers> patientIds = new HashSet<>();
+        int bundleNo = 1;
+        try {
+
+
```
Comment (nflearl) [inline worker/src/main/java/gov/cms/ab2d/worker/processor/CoverageMappingCallable.java:68]: trim

File: worker/src/main/java/gov/cms/ab2d/worker/processor/CoverageMappingCallable.java [hunk 11]
```diff
@@ -0,0 +1,296 @@
...
+                    if (bundle.getLink(Bundle.LINK_NEXT) == null) {
+                        log.info("retrieving contract membership for Contract {}-{}-{} bundle #{}, does not have a next link",
+                                contractNumber, year, month, bundleNo);
+                    }
+
+                    patientIds.addAll(extractAndFilter(bundle));
+                }
+            } catch (InternalErrorException ie) {
+                // Catch edge case bug where (number patients) mod (bundle size) == 0
+                // Extra bundle link returned that has no data in it which causes exception
```
Comment (nflearl) [inline worker/src/main/java/gov/cms/ab2d/worker/processor/CoverageMappingCallable.java:114]: This should be fixed by the changes made by me, Christopher and Karl.  Please re-verify and remove if no longer needed.

The same goes for PatientContractCallable.  In fact, there may be an opportunity to create an abstract base class for these two classes and do a pull up refactoring of common functionality.

**Review by nflearl — CHANGES_REQUESTED** [review]
With the Paging fixes in BFD, need to retest and potentially remove some work arounds.

Mostly nits and requests for addition of TODO comments.

Also a modest refactoring or two.

File: worker/src/main/java/gov/cms/ab2d/worker/processor/CoverageMappingCallable.java (hunk 9 above)
Comment (wnyffenegger) [reply worker/src/main/java/gov/cms/ab2d/worker/processor/CoverageMappingCallable.java:27]: The PatientContractCallable will be deleted in the future and replaced by this class because we are loading the data into our database beforehand. Not sure if we should make a class when the intention is for the PatientContractCallable to be deleted in the near future.

File: worker/src/main/java/gov/cms/ab2d/worker/processor/CoverageMappingCallable.java (hunk 11 above)
Comment (wnyffenegger) [reply worker/src/main/java/gov/cms/ab2d/worker/processor/CoverageMappingCallable.java:114]: I'll check whether the fix works. As for the rest the goal is to remove the PatientContractCallable completely so I'm not sure refactoring is a great idea.

File: Deploy/bash/set-env.sh [hunk 12]
```diff
@@ -19,17 +19,17 @@ INVOKED_COMMAND="$(printf %q "$BASH_SOURCE")$((($#)) && printf ' %q' "$@")"
...
-  echo "*********************************************"
-  echo "ERROR: This script must be run using 'source'"
-  echo "---------------------------------------------"
-  echo "EXPECTED FORMAT:"
-  echo "source ${INVOKED_COMMAND}"
-  echo "*********************************************"
-  echo ""
-  exit 1
-fi
+#if [ "$0" != "-bash" ]; then
```
[the remark this answers is not in the document]
Comment (wnyffenegger) [reply Deploy/bash/set-env.sh:22]: Commenting that code out should not be in this PR and needs to be removed.

File: common/src/main/java/gov/cms/ab2d/common/model/CoverageSearch.java [hunk 13]
```diff
@@ -0,0 +1,36 @@
...
+    private Long id;
+
+    @OneToOne
+    @JoinColumn(name = "bene_coverage_period_id", referencedColumnName = "id")
+    private CoveragePeriod period;
+
+    // We can use this to search for the earliest search request
+    private OffsetDateTime created;
+
+    private int attempts;
```
[the remark this answers is not in the document]
Comment (wnyffenegger) [reply common/src/main/java/gov/cms/ab2d/common/model/CoverageSearch.java:26]: Commented. This is used in a circuit breaker which trips when the number of failed attempts made to retrieve membership for a coverage period has failed after x attempts.

When a search fails against BFD this number is incremented.

File: common/src/main/java/gov/cms/ab2d/common/repository/CoveragePeriodRepository.java [hunk 14]
```diff
@@ -11,5 +12,9 @@
 
     List<CoveragePeriod> findAllByMonthAndYear(int month, int year);
 
+    List<CoveragePeriod> findAllByLastSuccessfulJobIsNull();
+
+    List<CoveragePeriod> findAllByMonthAndYearAndLastSuccessfulJobLessThanEqual(int month, int year, OffsetDateTime time);
```
[the remark this answers is not in the document]
Comment (wnyffenegger) [reply common/src/main/java/gov/cms/ab2d/common/repository/CoveragePeriodRepository.java:17]: Yeah well I don't have to write the sql for it so that's something.

File: worker/src/main/java/gov/cms/ab2d/worker/processor/CoverageProcessor.java [hunk 15]
```diff
@@ -0,0 +1,44 @@
...
+     */
+    void queueStaleCoveragePeriods();
+
+
+    /**
+     * Add a {@link CoveragePeriod} to the queue of periods to be mapped
+     * @param period coverage period to add
+     * @param prioritize if true insert in front of the queue
+     */
+    void queueCoveragePeriod(CoveragePeriod period, boolean prioritize);
```
[the remark this answers is not in the document]
Comment (wnyffenegger) [reply worker/src/main/java/gov/cms/ab2d/worker/processor/CoverageProcessor.java:24]: I think we should put this off to another PR. @lailasharshar you had some comments concerning prioritizing coverage membership searches.

File: worker/src/test/java/gov/cms/ab2d/worker/processor/CoverageMappingCallableTest.java [hunk 16]
```diff
@@ -0,0 +1,377 @@
...
+
+            assertEquals(10, missingIdentifier);
+        } catch (Exception exception) {
+            fail("could not execute basic job with mock client", exception);
+        }
+
+    }
+
+    /**
+     * Protects against an issue with BFD api where an edge case can cause
```
[the remark this answers is not in the document]
Comment (wnyffenegger) [reply worker/src/test/java/gov/cms/ab2d/worker/processor/CoverageMappingCallableTest.java:277]: No longer necessary since you resolved it with Chris

File: worker/src/main/java/gov/cms/ab2d/worker/processor/CoverageMappingCallable.java (hunk 11 above)
Comment (wnyffenegger) [reply worker/src/main/java/gov/cms/ab2d/worker/processor/CoverageMappingCallable.java:114]: The fix does work

File: worker/src/test/java/gov/cms/ab2d/worker/service/WorkerServiceDisengagementTest.java [hunk 17]
```diff
@@ -58,7 +58,7 @@
     public void init() {
         jobRepository.deleteAll();
         userRepository.deleteAll();
-        coverageRepository.deleteAll();
+        dataSetup.deleteCoverage();
```
[the remark this answers is not in the document]
Comment (wnyffenegger) [reply worker/src/test/java/gov/cms/ab2d/worker/service/WorkerServiceDisengagementTest.java:61]: Tried a quick fix and it broke things.

**Review by nflearl — APPROVED** [review]

**Review by lailasharshar — APPROVED** [review]

**Review by lailasharshar — DISMISSED** [review]

**Review by nflearl — APPROVED** [review]

**Review by nflearl — APPROVED** [review]

**Review by lailasharshar — APPROVED** [review]
